File: backend/server/routers/collections.py
# ---------------------------------------------------------------------------
# system: ModelArchivist
# file: collections.py
# purpose: REST interface for collections
# ---------------------------------------------------------------------------

import backend.repository.repository as repo
from fastapi import APIRouter, HTTPException, Response
from pydantic import BaseModel

from backend.exception import ArcException
from backend.dispatcher import OperationBusyError, dispatcher
from backend.repository.tables import DeploymentStatus

router = APIRouter()


class CollectionModelUpdate(BaseModel):
    model_ids: list[str]
    add: bool


class CollectionWorkflowUpdate(BaseModel):
    workflow_ids: list[str]
    add: bool


class CollectionUserObjectUpdate(BaseModel):
    user_object_ids: list[str]
    add: bool

@router.get('/collections')
async def get_collections() -> list[dict]:
    return repo.list_collections(True)


@router.get('/collections/{id}')
async def get_collection(id: str) -> dict:
    try:
        return repo.get_collection(id)
    except ArcException as error:
        if error.code == ArcException.Code.UNKNOWN_COLLECTION:
            raise HTTPException(404, error.message)
        raise HTTPException(400, error.message)


@router.post('/collections')
async def create_collection(data: dict) -> dict:
    try:
        return repo.create_collection(data)
    except ArcException as error:
        if error.code in (ArcException.Code.UNKNOWN_MODEL,
                          ArcException.Code.UNKNOWN_WORKFLOW,
                          ArcException.Code.UNKNOWN_USER_OBJECT,
                          ArcException.Code.UNKNOWN_COLLECTION):
            raise HTTPException(404, error.message)
        if error.code in (ArcException.Code.DUPLICATE_COLLECTION_MEMBER,
                          ArcException.Code.COLLECTION_CYCLE):
            raise HTTPException(409, error.message)
        raise HTTPException(400, error.message)


@router.put('/collections/{id}')
async def update_collection(id: str, data: dict) -> dict:
    try:
        return repo.update_collection(id, data)
    except ArcException as error:
        if error.code in (ArcException.Code.UNKNOWN_MODEL,
                          ArcException.Code.UNKNOWN_WORKFLOW,
                          ArcException.Code.UNKNOWN_USER_OBJECT,
                          ArcException.Code.UNKNOWN_COLLECTION):
            raise HTTPException(404, error.message)
        if error.code in (ArcException.Code.DUPLICATE_COLLECTION_MEMBER,
                          ArcException.Code.COLLECTION_CYCLE):
            raise HTTPException(409, error.message)
        raise HTTPException(400, error.message)


@router.post('/collections/{id}/models')
async def update_collection_models(id: str, data: CollectionModelUpdate) -> dict:
    try:
        return repo.update_collection_models(id, data.model_ids, data.add)
    except ArcException as error:
        if error.code in (ArcException.Code.UNKNOWN_MODEL,
                          ArcException.Code.UNKNOWN_COLLECTION):
            raise HTTPException(404, error.message)
        if error.code in (ArcException.Code.EMPTY_COLLECTION,
                          ArcException.Code.DUPLICATE_COLLECTION_MEMBER):
            raise HTTPException(409, error.message)
        raise HTTPException(400, error.message)


@router.post('/collections/{id}/workflows')
async def update_collection_workflows(id: str, data: CollectionWorkflowUpdate) -> dict:
    try:
        return repo.update_collection_workflows(id, data.workflow_ids, data.add)
    except ArcException as error:
        if error.code in (ArcException.Code.UNKNOWN_WORKFLOW,
                          ArcException.Code.UNKNOWN_COLLECTION):
            raise HTTPException(404, error.message)
        if error.code in (ArcException.Code.EMPTY_COLLECTION,
                          ArcException.Code.DUPLICATE_COLLECTION_MEMBER):
            raise HTTPException(409, error.message)
        raise HTTPException(400, error.message)


@router.post('/collections/{id}/user-objects')
async def update_collection_user_objects(id: str, data: CollectionUserObjectUpdate) -> dict:
    try:
        return repo.update_collection_user_objects(id, data.user_object_ids, data.add)
    except ArcException as error:
        if error.code in (ArcException.Code.UNKNOWN_USER_OBJECT,
                          ArcException.Code.UNKNOWN_COLLECTION):
            raise HTTPException(404, error.message)
        if error.code in (ArcException.Code.EMPTY_COLLECTION,
                          ArcException.Code.DUPLICATE_COLLECTION_MEMBER):
            raise HTTPException(409, error.message)
        raise HTTPException(400, error.message)


@router.delete('/collections/{id}')
async def delete_collection(id: str) -> dict:
    try:
        return repo.delete_collection(id)
    except ArcException as error:
        if error.code == ArcException.Code.UNKNOWN_COLLECTION:
            raise HTTPException(404, error.message)
        if error.code == ArcException.Code.EMPTY_COLLECTION:
            raise HTTPException(409, error.message)
        raise HTTPException(400, error.message)


@router.post('/collections/{id}/synchronize')
async def synchronize_collection(id: str, response: Response,
                                 simulate: bool = True) -> dict:
    plan = repo.synchronize_collection(id, True)
    if simulate or not plan['allowed']:
        return plan
    if not repo.collection_operation_requires_lro(plan):
        return repo.synchronize_collection(id, False)
    try:
        operation = dispatcher.submit(
            'collection_sync',
            lambda _report: repo.synchronize_collection(id, False))
    except OperationBusyError as error:
        raise HTTPException(409, str(error))
    response.status_code = 202
    return operation


@router.post('/collections/{id}/move')
async def move_collection(id: str, destination: DeploymentStatus,
                          response: Response, simulate: bool = True) -> dict:
    plan = repo.move_collection(id, destination, True)
    if simulate or not plan['allowed']:
        return plan
    if not repo.collection_operation_requires_lro(plan):
        return repo.move_collection(id, destination, False)
    try:
        operation = dispatcher.submit(
            'collection_move',
            lambda _report: repo.move_collection(id, destination, False))
    except OperationBusyError as error:
        raise HTTPException(409, str(error))
    response.status_code = 202
    return operation

File: frontend/src/lib/collections.ts
/* ---------------------------------------------------------------------------
 * system: ModelArchivist
 * file: frontend/src/lib/collections.ts
 * purpose: Collection handling
 * ---------------------------------------------------------------------------*/

import {
    type CollectionSummary,
    type CollectionOverview,
    type Collection,
    identity } from "$lib/objects";

import {
    type ApiResult,
    getUrl,
    parseResponse } from "$lib/api";

export type CollectionSearchCriteria = {
    types: string[];
    collections: string[];
    required_tags: string[];
    forbiddenTags: string[];
    name: string;
}

export type CollectionInput = {
    name: string;
    purpose: string;
    tags: string[];
    models: string[];
    workflows: string[];
    user_objects?: string[];
    children: string[];
};

export async function getCollections(): Promise<ApiResult<CollectionOverview[]>> {
    const url = getUrl('/collections');
    const response = await fetch(url);
    return await parseResponse(response, identity, 'getCollections');
}

export async function searchCollections(criteria: CollectionSearchCriteria): Promise<ApiResult<CollectionSummary[]>> {
    const url = getUrl('/collections/search');
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(criteria)
    });
    return await parseResponse(response, identity, 'searchCollections');
}

export async function getCollection(collectionId: string): Promise<ApiResult<Collection>> {
    const url = getUrl(`/collections/${collectionId}`);
    const response = await fetch(url)
    return await parseResponse(response, identity, 'getCollection');
}

export async function updateCollection(collectionId: string,
                                       updatedCollection: CollectionInput): Promise<ApiResult<CollectionSummary>> {
    const url = getUrl(`/collections/${collectionId}`);
    const response = await fetch(url, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updatedCollection)
    });
    return await parseResponse(response, identity, 'updateCollection')
}

export async function createCollection(collection: CollectionInput): Promise<ApiResult<CollectionSummary>> {
    const url = getUrl('/collections');
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(collection)
    });
    return await parseResponse(response, identity, 'createCollection');
}

export function collectionInput(collection: Collection): CollectionInput {
    return {
        name: collection.name,
        purpose: collection.purpose,
        tags: [...collection.tags],
        models: collection.models.map((model) => model.id),
        workflows: collection.workflows.map((workflow) => workflow.id),
        user_objects: collection.user_objects.map((item) => item.id),
        children: collection.children.map((child) => child.id)
    };
}

export type MemberField = 'models' | 'workflows' | 'user_objects' | 'children';
export type CollectionMember = {id: string; name: string; has_archive: boolean; has_working: boolean};
export type MemberSegment = {id: string; name: string; field: MemberField;
    members: CollectionMember[]; candidates: CollectionMember[]};
export type CollectionOperationResult = {allowed: boolean; performed?: boolean;
    errors?: {code: string; message: string}[]; warnings?: {code: string; message: string}[];
    members?: CollectionOperationResult[]};

export async function getCollectionMembers(id: string): Promise<ApiResult<MemberSegment[]>> {
    const response = await fetch(getUrl(`/collections/${id}/members`));
    return await parseResponse(response, identity, 'getCollectionMembers');
}

export async function operateCollection(id: string, destination: 'working' | 'archive' | null):
    Promise<ApiResult<import('$lib/models').Operation | CollectionOperationResult>> {
    const action = destination === null ? 'synchronize?simulate=false'
        : `move?simulate=false&destination=${destination}`;
    const response = await fetch(getUrl(`/collections/${id}/${action}`), {method: 'POST'});
    return await parseResponse(response, identity, 'operateCollection');
}

export async function addModelToCollection(collectionId: string,
                                           modelId: string): Promise<ApiResult<CollectionSummary>> {
    const current = await getCollection(collectionId);
    if (!current.ok) return current;
    const input = collectionInput(current.data);
    input.models.push(modelId);
    return await updateCollection(collectionId, input);
}

export async function removeModelFromCollection(collectionId: string,
                                                modelId: string): Promise<ApiResult<CollectionSummary>> {
    const current = await getCollection(collectionId);
    if (!current.ok) return current;
    const input = collectionInput(current.data);
    input.models = input.models.filter((id) => id !== modelId);
    return await updateCollection(collectionId, input);
}

export async function updateCollectionModels(collectionId: string, modelIds: string[],
                                             add: boolean): Promise<ApiResult<CollectionSummary>> {
    const response = await fetch(getUrl(`/collections/${collectionId}/models`), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({model_ids: modelIds, add})
    });
    return await parseResponse(response, identity, 'updateCollectionModels');
}

export async function updateCollectionWorkflows(collectionId: string, workflowIds: string[],
                                                add: boolean): Promise<ApiResult<CollectionSummary>> {
    const response = await fetch(getUrl(`/collections/${collectionId}/workflows`), {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({workflow_ids: workflowIds, add})
    });
    return await parseResponse(response, identity, 'updateCollectionWorkflows');
}

export async function updateCollectionUserObjects(collectionId: string, userObjectIds: string[],
                                                  add: boolean): Promise<ApiResult<CollectionSummary>> {
    const response = await fetch(getUrl(`/collections/${collectionId}/user-objects`), {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({user_object_ids: userObjectIds, add})
    });
    return await parseResponse(response, identity, 'updateCollectionUserObjects');
}
